R/oauth-refresh.R

```diff
@@ -0,0 +1,113 @@
+# this file has its origins in oauth-refresh.R and oauth-error.R in httr
+# I want to introduce behaviour to error informatively for a deleted OAuth app
+
+# Refresh an OAuth 2.0 credential.
+#
+# Refreshes the given token, and returns a new credential with a
+# valid access_token. Based on:
+# https://developers.google.com/identity/protocols/oauth2/native-app#offline
+refresh_oauth2.0 <- function(endpoint, app, credentials, package = NULL) {
+  if (is.null(credentials$refresh_token)) {
+    abort("Refresh token not available")
+  }
+
+  refresh_url <- endpoint$access
+  req_params <- list(
+    refresh_token = credentials$refresh_token,
+    client_id = app$key,
+    client_secret = app$secret,
+    grant_type = "refresh_token"
+  )
+
+  response <- httr::POST(refresh_url, body = req_params, encode = "form")
+
+  err <- find_oauth2.0_error(response)
+  if (!is.null(err)) {
+    gargle_refresh_failure(err, app, package)
+    return(NULL)
+  }
+
+  httr::stop_for_status(response)
+  refresh_data <- httr::content(response)
+  utils::modifyList(credentials, refresh_data)
+}
+
+oauth2.0_error_codes <- c(
+  400,
+  401
+)
+
+# This implements error checking according to the OAuth2.0
+# specification: https://tools.ietf.org/html/rfc6749#section-5.2
+find_oauth2.0_error <- function(response) {
+  if (!httr::status_code(response) %in% oauth2.0_error_codes) {
+    return(NULL)
+  }
+
+  content <- httr::content(response)
+  if (is.null(content$error)) {
```

Comment:
This is more lax than httr, which checks against a fixed set of strings that, apparently, Google violates.

https://github.com/r-lib/httr/commit/ce8439f2d571e91bf846b43d5db3d89610568fa1

Reply:
That was probably written when I mistakenly believed that most APIs followed the OAuth spec.